rest/handler.go

```diff
@@ -412,32 +439,40 @@ func (h *handler) checkAuth(context *db.DatabaseContext) (err error) {
 	return nil
 }
 
-func (h *handler) checkAdminAuth(dbContext *db.DatabaseContext, basicAuthUsername, basicAuthPassword string, accessPermissions []string, responsePermissions []string) (statusCode int, err error) {
-	var managementEndpoints []string
-	var httpClient *http.Client
+func checkAdminAuth(bucketName, basicAuthUsername, basicAuthPassword string, httpClient *http.Client, managementEndpoints []string, accessPermissions []string, responsePermissions []string) (responsePermissionResults map[string]bool, statusCode int, err error) {
+	statusCode, permResults, err := CheckPermissions(httpClient, managementEndpoints, basicAuthUsername, basicAuthPassword, accessPermissions, responsePermissions)
+	if err != nil {
+		return nil, http.StatusInternalServerError, err
+	}
+
+	if len(responsePermissions) > 0 {
+		responsePermissionResults = permResults
+	}
 
-	if dbContext != nil {
-		managementEndpoints, httpClient, err = dbContext.ObtainManagementEndpointsAndHTTPClient()
-		if err != nil {
-			return http.StatusInternalServerError, err
-		}
-	} else {
-		managementEndpoints, httpClient, err = h.server.ObtainManagementEndpointsAndHTTPClient()
-		if err != nil {
-			return http.StatusInternalServerError, err
-		}
+	if statusCode == http.StatusOK {
+		return responsePermissionResults, http.StatusOK, nil
 	}
 
-	statusCode, permResults, err := h.server.CheckPermissions(httpClient, managementEndpoints, basicAuthUsername, basicAuthPassword, accessPermissions, responsePermissions)
-	if err != nil {
-		return http.StatusInternalServerError, err
+	// If status code was not 'ok' or 'forbidden' return an error
+	// If forbidden user has authenticated correctly but is not authorized via permissions. We'll fall through to try
+	// with roles.
+	if statusCode != http.StatusForbidden {
+		return nil, statusCode, nil
 	}
 
-	if len(responsePermissions) > 0 {
-		h.permissionsResults = permResults
+	var requestRoles []string
+	if bucketName != "" {
+		requestRoles = BucketScopedEndpointRoles[:]
+	} else {
+		requestRoles = ClusterScopedEndpointRoles[:]
+	}
+
+	statusCode, err = CheckRoles(httpClient, managementEndpoints, basicAuthUsername, basicAuthPassword, requestRoles, bucketName)
+	if err != nil {
+		return nil, http.StatusInternalServerError, err
 	}
 
-	return statusCode, nil
+	return permResults, statusCode, nil
```

Comment:
In the case where they have access through a role, this should also imply that they have access to anything otherwise gated by responsePermissions, shouldn't it?  I.e. set responsePermissionResults to match the incoming set, with all set to true?

Reply:
Yeah that's a good point.

Reply:
Yeah thats fair, added that case now.

Reply:
This actually raises a question... If say there is accessPermission A and responsePermissions B and C. And user has A, B and has role sync_gateway what would be the expected output?

Currently it'd find the user has access A, then just return with 'you have B responsePermission'. Should we fall through to do the role check anyway for that reason?

Reply:
As discussed,  in any case where the user fails a permission check (including responsePermissions), we need to fall through to role check before returning (for as long as we're supporting role-based authorization).